### PR TITLE
bug fix related to output sequence

### PR DESCRIPTION
**Pull request by raghavjajodia**
we should pick the last element of output sequence, because input is pre-padded with 0s, not post-padded

### CODE REVIEW

**Comment by Atcold** [comment]
@ritchieng, can you confirm? I think you coded this up.
I can look into it as well, if you don't remember / have time.

**Comment by Atcold** [comment]
So, in [`sequential_tasks.py:L4-L5`](https://github.com/Atcold/pytorch-Deep-Learning-Minicourse/blob/30266458b8eec0b2ce8f8355c7678d2300f85259/sequential_tasks.py#L4-L5) we have
```python
def pad_sequences(sequences, maxlen=None, dtype='int32', padding='pre', truncating='pre', value=0.):
```
and then, still in [`sequential_tasks.py:L207-L209`](https://github.com/Atcold/pytorch-Deep-Learning-Minicourse/blob/30266458b8eec0b2ce8f8355c7678d2300f85259/sequential_tasks.py#L207-L209) we have
```python
def encode_x_batch(self, x_batch):
    return pad_sequences([self.encode_x(x) for x in x_batch], maxlen=self.length_range[1])
```

Okay, so you wrote
```python
output = output[:, output.shape[1] - 1, :]
```
Can't we simply do
```python
output = output[:, -1, :]
```
instead?

**Comment by raghavjajodia** [comment]
Yes, I have already tested and verified all the outputs.
The reason why the old code works is that number of zero paddings are small. For example, 
for easy mode - padding = 0-2, max length = 9, seq length = 7-9
for normal mode - padding = 0-11, max length = 41, seq length = 30-41
for moderate mode - padding = 0-21, max length = 81, seq length = 60-81

and so on.. 
I tried an little extreme case
easy mode - padding = 43, max length = 50, seq length = 7
As expected, the old code terribly failed in learning anything, while the new code was still able to learn in this case.

**for example**
with input

The first element in the batch of sequences is:
 [[0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 1 0]
 [0 0 0 1 0 0 0 0]
 [1 0 0 0 0 0 0 0]
 [0 0 0 0 1 0 0 0]
 [1 0 0 0 0 0 0 0]
 [0 0 0 0 1 0 0 0]
 [0 0 0 1 0 0 0 0]
 [0 0 0 0 0 0 0 1]]

new code gives

Screen Shot 2019-07-12 at 10 03 01 AM

old code gives

Screen Shot 2019-07-12 at 10 09 21 AM

**Comment by raghavjajodia** [comment]
If you want to visualize the data and output variables, then following is an example.

data (input to model)
tensor([[0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 1., 0.],
        [0., 0., 0., 1., 0., 0., 0., 0.],
        [1., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 1., 0., 0., 0.],
        [1., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 1., 0., 0., 0.],
        [0., 0., 0., 1., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 1.]])

output - after calling "model(data)"

tensor([[-0.3018, -0.1259,  0.2750, -0.1177],
        [-0.2212, -0.1533,  0.3475, -0.2128],
        [-0.2397, -0.1738,  0.3159, -0.1789],
        [-0.2439, -0.1808,  0.3071, -0.1806],
        [-0.2144, -0.2737,  0.2796, -0.1810],
        [-0.2461, -0.2260,  0.2750, -0.1611],
        [-0.2110, -0.3031,  0.2636, -0.1743],
        [-0.2356, -0.2336,  0.2807, -0.1573],
        [-0.2084, -0.2537,  0.2982, -0.1885]], grad_fn=<SelectBackward>)

output after index selection - 

tensor([-0.2356, -0.2336,  0.2807, -0.1573], grad_fn=<SelectBackward>)

As we can see above, we should have picked the last vector, not second last vector

**Comment by Atcold** [comment]
Thank you. Plz, clean the output, and push one more commit.

**Comment by raghavjajodia** [comment]
Done. Thanks for maintaining these clean notebooks, I come back often here to refer many things, even after the DLSP19

**Comment by Atcold** [comment]
You are welcome, @raghavjajodia.
I'll add new notebook, next year, I hope.
For sure I need to include @ylecun energy based models (EBMs). Unless we play with them a little, I think it's hard to 'get' them, IMHO. Graph neural nets are missing too. Perhaps I should open an issue where we could list wanted material.

**Comment by raghavjajodia** [comment]
@Atcold  Yes, I 100% agree that we need EBM notebooks. Even I was thinking to write some generic EBM framework where we could try different architectures and loss functions (especially the ones mentioned by @ylecun in his slides). Sadly, I didn't get time to do this.

**Comment by Atcold** [comment]
Let me know if you'd like to create such notebook. We could work on it together.

**Comment by raghavjajodia** [comment]
I am down captain, lets do it!
